src/sync/treiber_stack.rs

```diff
@@ -49,6 +49,20 @@ impl<T> TreiberStack<T> {
         self.try_pop()
     }
 
+    /// Swaps a stack with a new one.  Can be used to take all the
+    /// elements off the stack at once.
+    ///
+    /// Wait-free
+    pub fn swap(&self, new: TreiberStack<T>) -> TreiberStack<T> {
+        let guard = epoch::pin();
+        // new is exclusively owned and so can be written and read
+        // with relaxed atomics
+        let newhead = new.head.swap(None, Relaxed, &guard);
```

Comment:
LLVM doesn't optimize it away in a minimal case, and I don't think LLVM will reliably trying to do such optimizations soon